File: Problem078-Subsets/Solution.java
/*
 * @SEU(master);@ZZULI(bachelor)
 * author:CuiXuan
 * email:873059043@qq.com
 * If you find any errors, please let me know.
 * If you have any better solution, please let me know.
 * 
 * Solution Using time 3ms
 */
public class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> returnColumn = new ArrayList<List<Integer>>();
        int[] tmp = new int[nums.length];
        Arrays.sort(nums);
        for(int i=0;i<=nums.length;i++){
            getSubset(returnColumn,tmp,0,-1,nums,i);
        }
        return returnColumn;
    }
    private void getSubset(List<List<Integer>> returnColumn,int[] tmp,int step,int index,int[] nums,int k){
        if(step==k){
            List<Integer> res = new ArrayList<Integer>();
            for(int i=0;i<k;i++)res.add(tmp[i]);
            returnColumn.add(res);
            return;
        }
        for(int i=index+1;i<=nums.length-k+step;i++){
            tmp[step]=nums[i];
            getSubset(returnColumn,tmp,step+1,i,nums,k);
        }
    }
}

File: Problem078-Subsets/Solution.cpp
/*
 * @SEU(master);@ZZULI(bachelor)
 * author:CuiXuan
 * email:873059043@qq.com
 * If you find any errors, please let me know.
 * If you have any better solution, please let me know.
 *
 * Solution Using time 8ms
 */
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> returnColumn;
        sort(nums.begin(),nums.end());
        vector<int> tmp(nums.size());
        for(int i=0;i<=nums.size();i++){
            getSubset(returnColumn,tmp,0,-1,nums,i);
        }
        return returnColumn;
    }
private:
    void getSubset(vector<vector<int>>& returnColumn,vector<int>& tmp,int step,int index,vector<int> nums,int k){
        if(step==k){
            vector<int> res(tmp.begin(),tmp.begin()+k);
            returnColumn.push_back(res);
            return;
        }
        for(int i=index+1;i<=nums.size()-k+step;i++){
            tmp[step]=nums[i];
            getSubset(returnColumn,tmp,step+1,i,nums,k);
        }
    }
};
